### PR TITLE
fix for ServerSentEvent.data being a multi-line string -- issue 4865

### PR DESCRIPTION
**Pull request by fizzy33**

### CODE REVIEW

File: core/src/main/scala/org/http4s/ServerSentEvent.scala [hunk 1]
```diff
@@ -30,7 +30,9 @@ final case class ServerSentEvent(
     retry: Option[Long] = None
 ) extends Renderable {
   def render(writer: Writer): writer.type = {
-    writer << "data: " << data << "\n"
+    data.linesIterator.foreach { dataLine =>
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/ServerSentEvent.scala:33]: I brought this up on the other ticket.  `linesIterator` will split on `\r`, `\r\n`, and `\n`.  This renderer will reassemble it with just `\n`, so it's a lossy conversion.  It's legal to separate the lines with `\r` or `\r\n`, so it doesn't _have_ to be lossy.

This is probably an exotic case, and what you've done is inarguably an improvement for real-world use.

File: tests/src/test/scala/org/http4s/ServerSentEventSpec.scala [hunk 2]
```diff
@@ -117,6 +121,11 @@ class ServerSentEventSpec extends Http4sSuite {
         .through(ServerSentEvent.decoder)
         .compile
         .toVector
+      val f = roundTrip.unsafeToFuture()
+      val rt = Await.result(f, Duration(10, TimeUnit.SECONDS))
+      if ( rt != sses ) {
+        toString
+      }
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/ServerSentEventSpec.scala:128]: I'm confused what's going on here, particularly with the blocking.  Is this for a better failure message?

**Review by rossabaker — COMMENTED** [review]
We can cherry-pick this to series/0.22.  The newer model described in #4865 could even go into series/0.22, if it beats our final release.  But don't worry about those logistics: we can cherry-pick it when the time comes.

File: tests/src/test/scala/org/http4s/ServerSentEventSpec.scala (hunk 2 above)
Comment (fizzy33) [reply tests/src/test/scala/org/http4s/ServerSentEventSpec.scala:128]: Ooof I will remove that is me hacking in something for debugging...

Comment (fizzy33) [reply tests/src/test/scala/org/http4s/ServerSentEventSpec.scala:128]: That code should have been removed.

**Comment by fizzy33** [comment]
moving the linesIterator talk here.

So switching to using this in render
```
    data.linesWithSeparators.foreach { dataLine =>
      writer << "data: " << dataLine
    }
```

Causes a bunch of regressions because the decoder uses `fs2.text.lines`. i.e. the decoder doesn't preserve line endings.  

So happy to chase that further if we think it is worth it.  Which one way I see would be have a 
```
  def linesWithEnding[F[_]]: Pipe[F, String, (String,Option[String])] = {
```
That the decoder uses.  Noting the Option[String] ending could interestingly be an enum CR, LF, CRLF or String (liking having a special type since it makes the API clear what is what.

**Comment by rossabaker** [comment]
A sum type for the line ending is fine with me, but I don't know what other "line endings" exist in exotic parts of Unicode if this is a general purpose `Pipe`.  Or might some other protocols consider a form feed character?  I think a sum type binds that pipe closely to SSE.

If that fs2 side trek is easy, let's do it.  If not, I don't think the line endings are nearly as much a real world concern as the multi-line rendering, or comments, and I'd be happy to clean that up later.  But if we punt on it, the property test should normalize carriage returns to `\n`, or it will be flaky.

**Comment by rossabaker** [comment]
0.22 and 0.23 release candidates are imminent, probably as soon as Circe is released.  If we want to get comment support, we'll need to add it to the model for bincompat reasons, or else it will have to wait for 1.0.  We might want to prioritize that over the carriage return preservation.

**Comment by rossabaker** [comment]
I think this is all included in #4888?

**Comment by fizzy33** [comment]
Agreed #4865 covers the multi-line data (and now comment as well).

**Comment by rossabaker** [comment]
I'll leave the original open to track the `\r` nuance, but we'll close this.

**Pull request closed without merging**